### PR TITLE
server: refactor handling of cleanup funcs

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind cleanup

> /kind dependency-change
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

#### What this PR does / why we need it:

it doesn't make very much sense to have so many deferred funcs queued, and check retErr each time
instead, we can check retErr once, and loop through a slice of cleanupFuncs

this also carries https://github.com/cri-o/cri-o/pull/4240 to avoid a merge conflict
#### Does this PR introduce a user-facing change?

```release-note
none
```

### CODE REVIEW

**Comment by haircommander** [comment]
pulled in https://github.com/cri-o/cri-o/pull/4198 to make sure that deferred func is included

File: server/container_create_linux.go [hunk 1]
```diff
@@ -372,6 +372,15 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrIface.Contai
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if retErr != nil {
+			log.Infof(ctx, "createCtrLinux: deleting container %s from storage", containerInfo.ID)
+			err2 := s.StorageRuntimeServer().DeleteContainer(containerInfo.ID)
+			if err2 != nil {
```
Comment (kolyshkin) [inline server/container_create_linux.go:379]:
```suggestion
			err := s.StorageRuntimeServer().DeleteContainer(containerInfo.ID); if err != nil {
```

File: server/container_create.go [hunk 2]
```diff
@@ -548,6 +548,14 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
 	if err := s.createContainerPlatform(newContainer, sb.CgroupParent(), mappings); err != nil {
 		return nil, err
 	}
+	defer func() {
+		if retErr != nil {
+			log.Infof(ctx, "createCtr: removing container ID %s from runtime", ctr.ID())
```
Comment (kolyshkin) [inline server/container_create.go:556]:
```suggestion
			log.Infof(ctx, "createContainer: removing container ID %s from runtime", ctr.ID())
```

File: server/container_create.go [hunk 3]
```diff
@@ -548,6 +548,14 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
...
+			if err2 := s.Runtime().DeleteContainer(newContainer); err2 != nil {
```
Comment (kolyshkin) [inline server/container_create.go:557]: doesn't have to be `err2`.

File: server/container_create_linux.go [hunk 4]
```diff
@@ -372,6 +372,15 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrIface.Contai
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if retErr != nil {
+			log.Infof(ctx, "createCtrLinux: deleting container %s from storage", containerInfo.ID)
```
Comment (kolyshkin) [inline server/container_create_linux.go:377]: the function is named `createSandboxContainer`

File: server/container_create.go (hunk 2 above)
Comment (kolyshkin) [reply server/container_create.go:556]: feel free to ignore :)

**Comment by kolyshkin** [comment]
I don't like the last commit much, it seems it complicates things (and decreases visibility) more than improves anything. `defer` is cheap in Go 1.15+, `!= nil` check is cheap, too.

**Comment by haircommander** [comment]
> I don't like the last commit much, it seems it complicates things (and decreases visibility) more than improves anything. `defer` is cheap in Go 1.15+, `!= nil` check is cheap, too.

Yeah, it's cheap, but doing it so many times is not needed. The main purpose of this PR is preparing for a larger refactor that requires we *not* cleanup when the error is not a context timeout. That'll add two extra checks: `err != context.Canceled && err != context.DeadlineExceeded` along with the non-nil check, which I think better justifies the PR.

File: internal/lib/sandbox/sandbox.go [hunk 5]
```diff
@@ -348,6 +348,14 @@ func (s *Sandbox) SetNetworkStopped(createFile bool) error {
 
 func (s *Sandbox) createFileInInfraDir(filename string) error {
 	infra := s.InfraContainer()
+	// If the sandbox is not yet created,
+	// this function is being called when
+	// cleaning up a failed sandbox creation.
+	// We don't need to create the file, as there will be no
+	// sandbox to restore
+	if !s.created {
```
Comment (fidencio) [inline internal/lib/sandbox/sandbox.go:356]: Isn't this the same case as pointed out by @kolyshkin here: https://github.com/cri-o/cri-o/pull/4258#pullrequestreview-506918116 ?

**Review by fidencio — COMMENTED** [review]
LGTM, there's a nitpick comment, but feel free to ignore it.

File: internal/lib/sandbox/sandbox.go (hunk 5 above)
Comment (haircommander) [reply internal/lib/sandbox/sandbox.go:356]: I'll rebase on https://github.com/cri-o/cri-o/pull/4241 once it merges :)

**Comment by haircommander** [comment]
this is ready @fidencio @umohnani8 @mrunalp @kolyshkin @giuseppe @saschagrunert

File: server/container_create.go [hunk 6]
```diff
@@ -476,50 +476,53 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
...
-			log.Infof(ctx, "createCtr: deleting container %s from storage", ctr.ID())
-			err2 := s.StorageRuntimeServer().DeleteContainer(ctr.ID())
-			if err2 != nil {
-				log.Warnf(ctx, "Failed to cleanup container directory: %v", err2)
-			}
+	cleanupFuncs = append(cleanupFuncs, func() {
+		log.Infof(ctx, "createCtr: deleting container %s from storage", ctr.ID())
+		err2 := s.StorageRuntimeServer().DeleteContainer(ctr.ID())
+		if err2 != nil {
+			log.Warnf(ctx, "Failed to cleanup container directory: %v", err2)
```
Comment (TomSweeneyRedHat) [inline server/container_create.go:507]: soft suggest "directory" -> "storage"

File: server/sandbox_run_linux.go [hunk 7]
```diff
@@ -535,14 +539,12 @@ func (s *Server) runPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
...
-		defer func() {
-			if retErr != nil {
-				log.Infof(ctx, "runSandbox: unmounting shmPath for sandbox %s", sbox.ID())
-				if err2 := unix.Unmount(shmPath, unix.MNT_DETACH); err2 != nil {
-					log.Warnf(ctx, "failed to unmount shm for pod: %v", err2)
-				}
+		cleanupFuncs = append(cleanupFuncs, func() {
+			log.Infof(ctx, "runSandbox: unmounting shmPath for sandbox %s", sbox.ID())
+			if err2 := unix.Unmount(shmPath, unix.MNT_DETACH); err2 != nil {
+				log.Warnf(ctx, "failed to unmount shm for pod: %v", err2)
```
Comment (TomSweeneyRedHat) [inline server/sandbox_run_linux.go:545]: you say "for pod" here, but "for sandbox" at line 543?  I'd be consistent.

File: server/sandbox_run_linux.go [hunk 8]
```diff
@@ -656,27 +656,23 @@ func (s *Server) runPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
...
-	defer func() {
-		if retErr != nil {
-			log.Infof(ctx, "runSandbox: removing pod sandbox %s", sbox.ID())
-			if err := s.removeSandbox(sbox.ID()); err != nil {
-				log.Warnf(ctx, "could not remove pod sandbox: %v", err)
-			}
+	cleanupFuncs = append(cleanupFuncs, func() {
+		log.Infof(ctx, "runSandbox: removing pod sandbox %s", sbox.ID())
+		if err := s.removeSandbox(sbox.ID()); err != nil {
+			log.Warnf(ctx, "could not remove pod sandbox: %v", err)
```
Comment (TomSweeneyRedHat) [inline server/sandbox_run_linux.go:662]: You have "could not" here, and "couldn't" in a number of other messages.  I'd pref "could not" through out.

File: server/sandbox_run_linux.go [hunk 9]
```diff
@@ -656,27 +656,23 @@ func (s *Server) runPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
...
-	defer func() {
-		if retErr != nil {
-			log.Infof(ctx, "runSandbox: deleting pod ID %s from idIndex", sbox.ID())
-			if err := s.PodIDIndex().Delete(sbox.ID()); err != nil {
-				log.Warnf(ctx, "couldn't delete pod id %s from idIndex", sbox.ID())
-			}
+	cleanupFuncs = append(cleanupFuncs, func() {
+		log.Infof(ctx, "runSandbox: deleting pod ID %s from idIndex", sbox.ID())
+		if err := s.PodIDIndex().Delete(sbox.ID()); err != nil {
+			log.Warnf(ctx, "couldn't delete pod id %s from idIndex", sbox.ID())
```
Comment (TomSweeneyRedHat) [inline server/sandbox_run_linux.go:673]:
```suggestion
			log.Warnf(ctx, "couldn't delete pod ID %s from idIndex", sbox.ID())
```

File: server/sandbox_run_linux.go [hunk 10]
```diff
@@ -938,14 +924,12 @@ func (s *Server) runPodSandbox(ctx context.Context, req *pb.RunPodSandboxRequest
...
 			return nil, err
 		}
-		defer func() {
-			if retErr != nil {
-				log.Infof(ctx, "runSandbox: in not manageNSLifecycle, stopping network for sandbox %s", sb.ID())
-				if err2 := s.networkStop(ctx, sb); err2 != nil {
-					log.Errorf(ctx, "error stopping network on cleanup: %v", err2)
-				}
+		cleanupFuncs = append(cleanupFuncs, func() {
+			log.Infof(ctx, "runSandbox: in not manageNSLifecycle, stopping network for sandbox %s", sb.ID())
```
Comment (TomSweeneyRedHat) [inline server/sandbox_run_linux.go:928]: Perhaps?
```suggestion
			log.Infof(ctx, "runSandbox: is not in manageNSLifecycle, stopping network for sandbox %s", sb.ID())
```

**Comment by TomSweeneyRedHat** [comment]
done picking.  Just a few suggestions for consideration, only message near the end that was badly formed and should really be reworked.

**Comment by haircommander** [comment]
nits addressed!

File: server/container_create.go [hunk 11]
```diff
@@ -476,50 +476,53 @@ func (s *Server) CreateContainer(ctx context.Context, req *pb.CreateContainerReq
 		return nil, errors.Wrap(err, "setting container name and ID")
 	}
 
+	cleanupFuncs := make([]func(), 0)
```
Comment (kolyshkin) [inline server/container_create.go:479]: I think `, 0` is redundant

Comment (haircommander) [reply server/container_create.go:479]: you need to specify a length when initializing a slice

Comment (kolyshkin) [reply server/container_create.go:479]: Right, I always forget about this.

```go
cleanupFuncs := []func(){}
```
maybe?

**Comment by kolyshkin** [comment]
Side note: there is a way that looks nice, does not require `defer`s or repeated error checks. It only requires labels and goto.  
Something like this (pseudocode):

```go
var retErr error

err: = do_a(); if err != nil {
   retErr = errors.Wrap(err, "failed to do a")
   goto undo_a
}

if err := do_b(); err != nil {
   retErr = errors.Wrap(err, "failed to do b")
   goto undo_b
}

if err := do_c(); err != nil {
   retErr = errors.Wrap(err, "failed to do c")
   goto undo_c
}

return nil

undo_c:
   undo_c()
undo_b:
   undo_b()
undo_a:
   undo_a()
log(retErr)
return retErr
```

Surely, wrapping and logging are optional

File: server/container_create.go (hunk 11 above)
Comment (haircommander) [reply server/container_create.go:479]: is there any reason that is better than using make (which feels more idiomatic to me)

**Comment by haircommander** [comment]
> Side note: there is a way that looks nice, does not require defers, named return values, or repeated error checks. It only requires labels and goto.
> ....
> Surely, wrapping and logging are optional. Basically it is like lots of nested `if`s, but without the ugliness of too much nesting.

I'm not sure I want to go down this route, I'm not crazy about `goto`s and I don't think it really cleans up the code all that much. Right now, this change is pretty small and will be useful later, are there any blocking comments left

@kolyshkin @giuseppe @mrunalp @umohnani8 @saschagrunert @TomSweeneyRedHat

**Review by saschagrunert — APPROVED** [review]
LGTM

**Comment by haircommander** [comment]
PTAL @umohnani8 @mrunalp @kolyshkin @saschagrunert

**Comment by haircommander** [comment]
e2e-gcp passed, force merging instead of wasting more cycles on flakes

**Pull request merged**